aip/0194.md

```diff
@@ -0,0 +1,237 @@
+  other reasons than quota and the expected time for the resource to become
+  available is much shorter, it may be retryable.
+- `NOT_FOUND`: The resource may be created in the future.
+- `ALREADY_EXISTS`: The resource may be deleted in the future.
+- `PERMISSION_DENIED`: The user may be authenticated in the future.
+- `FAILED_PRECONDITION`: The system state may change in the future.
+- `OUT_OF_RANGE`: The range may be extended in the future.
+- `UNIMPLEMENTED`: The RPC may be implemented in the future.
+- `UNKNOWN`: `UNKNOWN` and `INTERNAL` are both generic error codes - they are
+  errors that may be generated from a wide breadth of causes. Known services
+  within the system **should** generate `INTERNAL` errors when the error doesn't
+  fit any other code. `UNKNOWN` errors are reserved for truly
+  unknown-to-the-system errors, and therefore **should** immediately surfaced to
+  the user.
+
+## Appendix
+
+### Idempotency and pragmaticism
+
+A common concern with retrying `CREATE` and `DELETE` operations is that they are
+non-idempotent operations. The functional concern can be summarized with this
+example:
+
+```
+Get("Foo")
+> NOT_FOUND
+Create("Foo") // Let's work under the assumption that this should be retried.
+// ----> UNAVAILABLE
+// ----> Create("Foo") # retry #1
+// ----> UNAVAILABLE   # actually, the request succeeded, but the connection died during the response
+// ----> Create("Foo") # retry #2
+// ----> ALREADY_EXISTS
+> ALREADY_EXISTS
+```
+
+Here, an unexpected result - `ALREADY_EXISTS` - is returned from a Create. The
+resource didn't exist before the call was initiated, so the user may wonder,
+"How did I receive this result?".
+
+Does this unexpected result make the call non-idempotent?
+
+The definition of idempotency is unfortunately vague in this regard. [Oxford
+dictionary defines idempotent as](https://www.lexico.com/en/definition/idempotent),
+
+> denoting an element of a set which is unchanged in value when multiplied or
+> otherwise operated on by itself.
+
+One interpretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _response_.
+
+This definition allows for retrying `READ` and `UPDATE` requests, since
+repeated runs always result in the same state and response code. However, this
+definition precludes retrying `CREATE` calls, since `ALREADY_EXISTS` may be
+returned unexpectedly (as in the above example). It also precludes retrying
+`DELETE` calls, since `NOT_FOUND` may be returned unexpectedly.
+
+Another intepretation of this definition is that an operation is idempotent if
+repeating the operation results in the same _state_ (or, value).
+
+This second definition allows for retrying all CRUD operations, since the
+interpretation is only concerned with state, and repeating any CRUD operation
+always results in the same state.
+
+This document's retryability guidances follows the latter interpretation,
```

Comment:
Worth noting: I still disagree with this, and I think @jgeewax does too.

There are some other problem cases also. Let's take a modification of your example with _two_ actors:
```
> Get("Foo")
< NOT_FOUND

# Actor A
> Create("Foo")
< UNAVAILABLE  # it actually did not work

# Actor B
> Create("Foo")
< OK

# Actor A
> Create("Foo")  # retry 1
< ALREADY_EXISTS   # correct! and not A's.
```

That is semantically different in an important way. To put the difference in perspective, let's say than an API implements the [AIP-154](https://aip.dev/154) `request_id` pattern -- users get _different_ results in these two cases. In your example, the user gets the resource. In my example, the user gets `ALREADY_EXISTS` because they lost the race.

Reply:
I'm having trouble parsing this.

>  In your example, the user gets the resource. In my example, the user gets ALREADY_EXISTS because they lost the race.

Is "user" Actor A?

>  To put the difference in perspective, let's say than an API implements the AIP-154 request_id pattern -- users get different results in these two cases. 

What are "these two cases"? Who is "users" - actor A or B?

--------------

The only difference in your example between actor A retrying the request and it not retrying the request is whether it would have had to have an `ALREADY_EXISTS` if statement or not. There is no downside to the user adding such a check.

Furthermore, if request_id is not implemented, in your example _both actors_ need such a check already. (If request_id is implemented, it might not be necessary (I'm not sure), but there's still no downside to having it)